src/jdk.jfr/share/classes/jdk/jfr/internal/tool/Metadata.java

```diff
@@ -127,16 +162,76 @@ public void execute(Deque<String> options) throws UserSyntaxException, UserDataE
-            try (RecordingFile rf = new RecordingFile(file)) {
-                List<Type> types = PRIVATE_ACCESS.readTypes(rf);
-                Collections.sort(types, new TypeComparator());
-                for (Type type : types) {
-                    prettyWriter.printType(type);
+
+            List<Type> types = null;
+            if (file != null) {
+                // has recording.jfr, read metadata from file
+                try (RecordingFile rf = new RecordingFile(file)) {
+                    types = JdkJfrConsumer.instance().readTypes(rf);
+                    Collections.sort(types, new TypeComparator());
                 }
+            } else {
+                // don't have recoring.jfr, read metadata from metadata.bin
+                types = TypeLibrary.getInstance().getTypes();
+            }
+
+            for (Type type : types) {
+                if (!(type instanceof PlatformEventType)) {
+                    continue;
+                }
+                if (foundEventFilter) {
+                    if (acceptedEvents.contains(type.getName())) {
+                        prettyWriter.printType(type);
+                        prettyWriter.flush(true);
+                    }
+                    continue;
+                }
+                if (foundCategoryFilter) {
+                    if (acceptCategory(acceptedCategories, type)) {
+                        prettyWriter.printType(type);
+                        prettyWriter.flush(true);
+                    }
+                    continue;
+                }
+                // no filter, just printing
+                prettyWriter.printType(type);
                 prettyWriter.flush(true);
-            } catch (IOException ioe) {
-                couldNotReadError(file, ioe);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static boolean acceptCategory(HashSet<String> categoryNames, Type type) {
+        Category categoryAnno = type.getAnnotation(Category.class);
+        int matchCount = 0;
+        if (categoryAnno != null) {
+            String[] categories = categoryAnno.value();
+            for (String category : categories) {
+                if (categoryNames.contains(category)) {
+                    matchCount++;
+                }
+            }
+        }
+
+        return matchCount == categoryNames.size();
+    }
+
+    private Path getOptionalJFRInputFile(Deque<String> options) {
+        if (!options.isEmpty()) {
+            String file = options.getLast();
+            if (!file.startsWith("--")) {
+                try {
+                    Path tmp = Paths.get(file).toAbsolutePath();
+                    ensureAccess(tmp);
+                    ensureJFRFile(tmp);
+                    options.removeLast();
+                    return tmp;
+                } catch (Exception e) {
+                    // ignored since recording file for jfr metadata is optional
```

Comment:
Inadequate error handling. If there is something wrong with the file (permission etc) I like to know it, and not get the event types for the current JVM.

Reply:
It would not swallow the exception now.